pkg/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fswatcher
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+)
+
+// FSWatcher watches for files' changes and errors.
+type FSWatcher interface {
+	WatchFiles(paths []string, onChange func(), log *zap.Logger) error
+	Close() error
+}
+
+// fsWatcherWrapper wraps the fsnotify.Watcher and implements FSWatcher.
+type fsWatcherWrapper struct {
+	fsnotifyWatcher *fsnotify.Watcher
+}
+
+// WatchFiles adds files' directories to the watcher, store each file's hashed content,
+// and watch for their changes and errors. If a file's hashed content changes, invoke onChange().
+func (f *fsWatcherWrapper) WatchFiles(paths []string, onChange func(), log *zap.Logger) error {
+	fileHashContentMap := make(map[string]string)
+	uniqueDirs := make(map[string]bool)
+
+	go func() error {
+		for {
+			select {
+			case event := <-f.fsnotifyWatcher.Events:
+				if event.Op&fsnotify.Create == fsnotify.Create ||
+					event.Op&fsnotify.Write == fsnotify.Write ||
+					event.Op&fsnotify.Rename == fsnotify.Rename {
+					ok, newHash := isModified(event.Name, fileHashContentMap[event.Name])
```

Comment:
I am not sure you want to rely on the `event.Name` here, since you're watching parent directories, not the actual files. Just loop through all files and check each of them for change.